### ISSUE TITLE
Option for piping the video out via yt-dlp to mpv:

### ISSUE THREAD

**Issue by tharrisone** [woodruffw/ff2mpv#82]
There are some serious caching issues in mpv due to ffmpeg. I've gone over every cache setting in mpv and found only minimum changes no matter what.

I'd like an option to play the video via yt-dlp and piping it out to mpv. I made a very simple alias myself for the terminal and it uses my bandwidth properly and caches the whole video in seconds. 

play() {
yt-dlp -o - "$1" | mpv  -
}

I have been using mpv for many years but only recently started watching youtube/invidious videos through ff2mpv which is great. 
Also it would be helpful to make the firefox users aware that you need to download your scripts. I didn't know I needed to do that until I went to your home page from the addon page.

**Comment by woodruffw** [comment]
> Also it would be helpful to make the firefox users aware that you need to download your scripts. I didn't know I needed to do that until I went to your home page from the addon page.

Where else do you think we should advertise that the native client is needed? It's currently displayed prominently in the README on AMO:

<img width="833" alt="Screen Shot 2022-04-26 at 9 58 54 PM" src="https://user-images.githubusercontent.com/3059210/165443832-d9d68847-25f8-4681-a36f-77d08261e8ce.png">

(That's not a rhetorical question! If you know other places we should mention it, please let us know!)

**Comment by woodruffw** [comment]
Re: the original issue topic: yes, this is something that's frustrating about how MPV has chosen to implement their `youtube-dl`/`yt-dlp` support. 

I can think of a few potential options for us:

1. Do nothing, and hope it gets better over time. This would be compatible with our general policy of not duplicating or patching over MPV or `youtube-dl`'s behavior, but this is a case where the default behavior is pretty bad and unconfigurable.
2. Change the native clients to do piping by default. This will have the intended effect, but it also complicates the native clients *and* introduces assumptions about which `youtube-dl`/`yt-dlp` program the user has installed (and whether we can execute it, versus merely being able to exec MPV).
3. Show users how to patch their own native clients locally.
4. Something else?

**Comment by tharrisone** [comment]
Changing the native client does complicate things, and showing users how to patch the client would keep things as they are and offer an option for some people. Btw.. I don't know how I missed that link to your native client.

**Comment by DanSM-5** [comment]
I have to thank you for the nice tip @tharrisone. I've been looking for a way to cache things faster in mpv.

@woodruffw Here is an idea:

- Add a shim file to call the client. This shim file can be updated with the install script, so it can add a flag for the native client (something like `-p` for pipe or similar).
- Use this flag in the client to alter the behavior. If present it will do the piping call, else work as usual.

Based on my knowledge of the python client it could be implemented by abstracting both ways to call mpv into separate functions and call them accordingly. Although I don't know how much complexity is to call a pipe process in python. A quick search shows [this](https://stackoverflow.com/questions/41238273/execute-shell-command-with-pipes-in-python). I don't know anything about ruby, so I can't say much about the ruby client.

**Comment by pc00per** [comment]
> play() {
yt-dlp -o - "$1" | mpv -
}

This unfortunately streams only 720p version of videos. Can't stream video only + audio only ones that are off high quality...

**Comment by woodruffw** [comment]
Just to make sure everybody knows: if you know what you're doing and know what you want, it's perfectly fine to patch the native client on your own machine!

We try very hard not to break the native client between releases (and will never do so without new major version), so you can generally assume that you won't need to re-patch it whenever the browser-side extension updates.

**Comment by pc00per** [comment]
I noticed that streaming video only has some good gains in the seeking performance. So I've found another better way to stream through mpv, is to get URIs of video + audio streams separately & do a concatenated play. 
It doesn't speed up caching, but will make seeking little bit snappier (caz that's what we all care about)
`mpv $(yt-dlp -f bestvideo --get-url '$1') --audio-file=$(yt-dlp -f bestaudio --get-url '$1')`

Lemme know if anyone gained seek improvements with this compared to traditional yt-dlp defaults.
